homeassistant/components/wirelesstag.py

```diff
@@ -0,0 +1,299 @@
+        _LOGGER.error("Unable to connect to wirelesstag.net service: %s",
+                      str(ex))
+        hass.components.persistent_notification.create(
+            "Error: {}<br />"
+            "Please restart hass after fixing this."
+            "".format(ex),
+            title=NOTIFICATION_TITLE,
+            notification_id=NOTIFICATION_ID)
+        return False
+
+    # listen to custom event
+    hass.bus.listen('update_tags', hass.data[DOMAIN].handle_update_tags_event)
+    hass.bus.listen('tag_binary_event', hass.data[DOMAIN].handle_binary_event)
+
+    return True
+
+
+class WirelessTagBaseSensor(Entity):
+    """Base class for HA implementation for Wireless Sensor Tag."""
+
+    def __init__(self, api, tag):
+        """Initialize a base sensor for Wireless Sensor Tag platform."""
+        self._api = api
+        self._tag = tag
+        self._uuid = self._tag.uuid
+        self.tag_id = self._tag.tag_id
+        self._name = self._tag.name
+        self._state = None
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return True
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    def define_name(self, new_name):
+        """Allow to set new name for base sensor implementation."""
+        self._name = new_name
+
+    @property
+    def principal_value(self):
+        """Return base value.
+
+        Subclasses need override based on type of sensor.
+        """
+        return 0
+
+    def updated_state_value(self):
+        """Default implementation formats princial value."""
+        return self.decorate_value(self.principal_value)
+
+    # pylint: disable=no-self-use
+    def decorate_value(self, value):
+        """Decorate input value to be well presented for end user."""
+        return '{:.1f}'.format(value)
+
+    def update(self):
+        """Update state."""
+        if not self.should_poll:
+            return
+
+        updated_tags = self._api.load_tags()
+        updated_tag = updated_tags[self._uuid]
+        if updated_tag is None:
+            _LOGGER.error('Unable to update tag: "%s"', self.name)
+            return
+
+        self._tag = updated_tag
+        self._state = self.updated_state_value()
+
+    def update_tag_info(self, event):
+        """Update sensor data on push notification event.
+
+        Subclasses must ovverride - handler of push notification.
+        """
+        pass
+
+    def on_binary_event(self, event):
+        """Update sensor data on binary event.
+
+        Applicable for binary sensors only.
+        Subclasses must ovverride - handler of push notification.
+        """
+        pass
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {
```

Comment:
@balloob Would you agree if I will remove only static attributes? i.e ATTR_TAG_HW_REVISION, ATTR_TAG_FW_VERSION and ATTR_TAG_TYPE are static. 

Others are very dynamic:
- ATTR_VOLTAGE - actual battery voltage - dynamically changing and can be helpful for end user to identify is battery (which is replaceable) needs to be replaced. (automated as well)
- ATTR_TAG_COMMENT - a user defined description of specific tag. can be changed by end user...
- ATTR_TAG_SIGNAL_STRAIGHT - obviously very helpful for end user -allows to diagnose many of issues (no signal etc) - end user is able to change frequency used and other hw settings using this parameter - again it is a subject for automation or say be a trigger for automation
- ATTR_TAG_BEEP_DURATION - user is able to change it - there are about 15 options how tag should beep. I think possibly to add `service` to beep from hass and to allow control beep duration from hass  - in future versions...
- ATTR_TAG_OUT_OF_RANGE - again - by nature - it is very dynamic and very helpful for end user to see if tag is out of range  - tag can be in vehicle or in bag of your kids playing in yard... you can setup notifications on in range, out of range....
- ATTR_TAG_POWER_CONSUMPTION - this attributes changes depending on location of device (it's ATTR_TAG_SIGNAL_STRAIGHT) and allow to determine what would be battery life... very recommended to be shown to end user  - i use it myself to make sure specific tag is located in right way

Reply:
Just because it's user configurable doesn't mean that it's not mainly static. So comment should go. Out of range should be used to control the `available` property. Beep duration is a configuration option, we shouldn't expose. 

Home Assistant goal is not to be become a full replacement UI for controlling and configuring wireless tags (or any other IoT device for that matter).

Reply:
ok ) comment will be removed. as `out of range` - i use `is alive` to control `available` as pointed in previous review round. Do you propose to drop `is alive` in favor of `out of range` or combine them? 
IMO `is alive` closer to `available`...

Reply:
comment and beep duration are removed.

Reply:
If the tag is out of range, how would you know if it's alive? I think that available should be a combination of both ?

Reply:
Good question ). Info is provided not by tag, but by `tag manager` (a device connected by ethernet to local network) which communicates to actual tags that could be quite distant (up to about 300+ meters). I'm not sure how tag manager provides `alive` different from `out of range`. It is not really difficult technically to combine them. However, I think that guys from `wirelesstags` defined these two properties as different states for a reason - for example - it could be `in range` but so low battery that make it `not alive`... So I think `is alive` is an umbrella state for availability ie - it seems to me reasonable to: 
- provide property `availability` mapped to `is alive`
- keep `out of range` reported as attribute of entity

Would you agree?

Reply:
as side note - tags can be used to record what environmental conditions where during parcels transportation (by putting tag into track). tags are battery operated and some of them have internal memory which is used to record temperature/humidity/light etc so tag can be alive but `out of range` of tag manager. data could be read and analyzed when tag became in range of specific tag manager. Also the memory could save was it `alive` or not during trip.